config/legacyconfig/config.go

```diff
@@ -12,6 +20,28 @@ type LegacyK0sAndK3s struct {
 	Storage                          Storage            `json:"storage,omitempty"`
 }
 
+func (c *LegacyK0sAndK3s) DecodeYAML(r io.Reader) error {
+	o, err := io.ReadAll(r)
+	if err != nil {
+		return err
+	}
+
+	j, err := yaml.YAMLToJSON(o)
+	if err != nil {
+		return err
+	}
+
+	dec := json.NewDecoder(bytes.NewReader(j))
+	dec.DisallowUnknownFields()
+
+	err = dec.Decode(c)
+	if err != nil {
+		return fmt.Errorf("%w: %w", config.ErrInvalidFileFormat, err)
+	}
+
+	return nil
```

Comment:
Given that this function exists three times (including the current config decoding), can we refactor this into a generic function and then call that generic function from those three?